Add 24 word BIP-39 seed phrase support

kellybyrd opened the pull request:
UPDATED PR DESCRIPTION ON 2021-05-24:
Change the --seed option to take a "type" bip39 or mobile

The "--seed mobile" option will behave exactly like the --seed option did. It will accept only 12 word phrases from the mobile app, verifying that the checksum is 0.

The "--seed bip39" option will accept valid 12 or 24 word BIP39 phrases verifying they have the correct BIP39 checksum.

There are probably a few places I could have factored out some repetitive code, but I chose to leave it as is so the PR wasn't as big and hopefully easier to see what the changes are.

Note that this does not add path derivation or anything else many people expect from a BIP39 phrase. It is a first step towards that.

NOTE: After looking at your github CI pipeline, I ran `cargo fmt --all` and  `cargo clippy  -- -Dclippy::all` which reformatted the definition of `gen_keypair()`. I assume that is wanted, but not sure.

kellybyrd commented on src/mnemonic/mod.rs:
```diff
@@ -34,25 +36,31 @@ pub fn mnemonic_to_entropy(words: Vec<String>) -> Result<[u8; 32]> {
-    }
+    let (entropy_bits, _checksum_bits) = bits.split_at(divider_index);
+
+    // TODO: Implement checksum checking, while still accounting for bug in the
```
I disabled this because valid BIP-39 mnemonics will fail this check and I'm new enough at Rust that I didn't immediately see a great way to verify the checksum (AFAICT, it's  take the sha256 of the entropy_bits and compared those to checksum_bits, but those need to be bits and not vector of "1" and "0")

kellybyrd commented on src/mnemonic/mod.rs:
```diff
@@ -15,11 +15,13 @@ fn get_wordlist(language: Language) -> WordList {
 pub fn mnemonic_to_entropy(words: Vec<String>) -> Result<[u8; 32]> {
-    if words.len() != 12 {
-        bail!("Invalid number of seed words");
+    if words.len() != 12 && words.len() != 24 {
```
This (and how I handle entropy_bytes below) are kind of a hack hard coded for only two options, not the full range of BIP-39 mnemonics. I considered being more flexible here and below, but thought quicker to PR was a better choice for my first attempt at Rust.

kellybyrd commented:
Tagging @madninja and @abhay as requested on Discord

kellybyrd commented:
I have seen any comments on this, is this good as is? Or does it need changes to be merged? I'll rebase on current master this weekend just to be sure it applies cleanly without breaking anything.

abhay commented:
I wonder if we should make it clearer to folks when they enter an app seed phrase or a bip39 seed phrase that they have to explicitly specify

JayKickliter commented:
I'd like this. Please take a look #78. It might make sense to build on top of it then close #78

kellybyrd commented:
> I wonder if we should make it clearer to folks when they enter an app seed phrase or a bip39 seed phrase that they have to explicitly specify

I thought about that (literally today when answering another question from Discord about someone picking their own phrase). since this PR as is is pretty light, would you want me to add something like:
 * `--mobile-app-seed`. Replaces `--seed`, behaves like `--seed` does now. 
 * `--bip32-seed`. New, would check for proper checksum, would support only 12 and 24 word phrases for now (because I'm lazy and those are easy to do given the current entropy code.
 
I would update the help text to be specific about what each is for. I generally don't like long cmdline options, but in this case I think it will head off user confusion.

kellybyrd commented:
> I'd like this. Please take a look #78. It might make sense to build on top of it then close #78

I guess they're in the same code, but I think either could go in without the other.

kellybyrd commented:
> I wonder if we should make it clearer to folks when they enter an app seed phrase or a bip39 seed phrase that they have to explicitly specify

@abhay 
I updated my PR (including the description). I deprecated `--seed` and replaced it with `--seed-mobile` and `--seed-bip39`. the mobile seed behaves exactly as --seed used to (enforcing checksum bits of all `0`)  The bip39 seed option accepts 12 and 24 word phrases and verifies the BIP39 checksum. 

I added two unit tests, for both 12 and 24 --seed-bip39 phrases.

Please suggest changes, or apply changes yourselves

abhay commented on src/cmd/create.rs:
```diff
@@ -31,9 +31,17 @@ pub struct Basic {
 
     #[structopt(long)]
-    /// Use space separated seed words to create the wallet
+    /// Deprecated: Use --seed_moble.
```
nit: s/_/-/

kellybyrd replied:
nice catch. Fixed the underscore and the misspelling of `mobile`

abhay reviewed (commented):
This looks good to me from a UX perspective. Would still like @madninja and/or @lthiery's thoughts before landing.

JayKickliter reviewed (approved):
Madninja should sign off on the CLI changes, but this looks good to me.

lthiery commented on src/cmd/create.rs:
```diff
@@ -86,8 +103,13 @@ impl Cmd {
-        let seed_words = if self.seed {
-            Some(get_seed_words()?)
+
+        if self.seed {
```
maybe add another bail around here to guard against: ```self.seed_mobile && self.seed_bip39```?

another options is to make `--seed mobile/bip39` an enum perhaps. Is there a another seed type down the line like bip42?

kellybyrd replied:
Ahh, so you mean rather than have distinct flags for each type like I do now, make `--seed` take an argument on the cmdline that is the type, keeping the current logic which does `get_seed_words` after the cmdline is parsed?  That would definitely age better if there are more types. 

If I can't make that work, I'll definitely add a guard both --seed-* flags being set.

lthiery replied:
Yea, you can do it pretty easily with an enum type and the EnumString macro to do the "FromStr" parsing from the command line:

```
use strum_macros::EnumString;

#[derive(Debug, EnumString)]
pub enum Seed {
    mobile,
    bip39, 
}
```

Then in the Opt struct, you put:
```
    #[structopt(long)]    
    /// Use space separated seed words to generate a wallet from seed
    seed: Seed
```

kellybyrd replied:
Ok, just force pushed changes to cmd/create.rs and cmd/mod.rs so that the UI is now
`--seed bip39|mobile` 

Thanks for the suggestions, @lthiery , I like this much better.

kellybyrd commented on src/cmd/create.rs:
```diff
@@ -31,15 +40,16 @@ pub struct Basic {
 
     #[structopt(long, default_value = KEYTYPE_ED25519_STR)]
-    /// The type of key to generate (ecc_compact/ed25519(.
+    /// The type of key to generate (ecc_compact/ed25519)
```
Not related to my change, but noticed it (and the similar one below) while making changes here.

kellybyrd commented on src/cmd/create.rs:
```diff
@@ -109,18 +119,17 @@ impl Basic {
-        } else {
-            None
+        let seed_words = match self.seed {
+            Some(seed_type) => Some(get_seed_words(seed_type == SeedType::Bip39)?),
```
Is there a better way to do this `== Enum::Value` in rust? I didn't wan `get_seed_words()` or `gen_keypair()` to have to know about the Enum type, I just wanted to pass a bool. But, this made me add several traits to the enum definition just to do this comparison. 

Please let me know if there is something better to do here.

lthiery replied:
I would pass the enum type through all of these functions instead of maintaining this bool. So `fn get_seed_words(seed_type: SeedType)`. Do the same for `mnemonic_to_entropy`. You'll see your `if` statements all become `matches` and it'll be much easier to read IMO.

kellybyrd replied:
Done and pushed. I like this much better!

lthiery replied:
![image](https://user-images.githubusercontent.com/1606640/119526835-9921a480-bd34-11eb-91c7-9e32937c9e22.png)

lthiery reviewed (approved):
Nice feature and nice work! Thanks for the contribution, @kellybyrd !

kellybyrd commented:
> Nice feature and nice work! Thanks for the contribution, @kellybyrd !

Thanks! Ok @madninja, let me know if you have any changes you'd like to see on this.

madninja commented on src/cmd/create.rs:
```diff
@@ -12,13 +12,22 @@ use std::{
     Sharded(Sharded),
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, EnumString)]
```
Why do you need Copy and Clone?

kellybyrd replied:
> This looks great! Just a few minor hygiene comments.
> 
> I like the use of strum, I hadn't seen that yet. It adds a little bloat that could be avoided with simple FromStr and Display implementations, but I think that's ok

@lthiery  suggested it, but I just found `--format` that does it more manually. I'll go either way. 

I needed Copy and Clone to do some if compares that I removed. not sure if I still do. I'll check.

madninja replied:
I can see Clone, but you hardly ever need to define Copy.

madninja replied:
And here's an example of a `Zone` enum which doesn't need all of heck to do it's work:

```rust
#[derive(Debug, Clone, PartialEq)]
pub enum Zone {
    Data,
    Config,
}

impl FromStr for Zone {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        match s.to_lowercase().as_str() {
            "data" => Ok(Self::Data),
            "config" => Ok(Self::Config),
            _ => Err(Error::invalid_address()),
        }
    }
}

impl fmt::Display for Zone {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Config => f.write_str("config"),
            Self::Data => f.write_str("data"),
        }
    }
}
```

kellybyrd replied:
Ahh, I needed Copy, and Clone because I wasn't borrowing correctly, because I'm new to Rust. fixed. No more Copy or Clone and now taking immutable refs where I need to.

madninja commented on src/cmd/create.rs:
```diff
@@ -141,10 +150,14 @@ impl Sharded {
     match seed_words {
         Some(words) => {
-            let entropy = mnemonic_to_entropy(words)?;
+            let entropy = mnemonic_to_entropy(words, seed_type.expect("mnemonic_to_entropy"))?;
```
Either remove the `Option` wrapper around SeedType if a SeedType is always expected, or return an error... I'm not a fan of panics

kellybyrd replied:
will do.

kellybyrd replied:
done. I'm not 100% happy with the way this turned out, I ended up matching on a tuple (TIL you can math on tuples in Rust), and left the default code path as a bail!, but...getting to that path wouldn't be a user error, it would be a coding mistake. Not sure if there's a better way to handle this, the only other thing I thought of was wrapping up the words and the type in its own small struct that is either all filled out or None. But, that felt even more complicated.

madninja commented on src/cmd/create.rs:
```diff
@@ -12,13 +12,22 @@ use std::{
 
+#[derive(Copy, Clone, Debug, PartialEq, EnumString)]
+#[strum(serialize_all = "mixed_case")]
+pub enum SeedType {
```
Move this definition to the mnemonic level? importing "cmd" types into non cmd code is not a good convention and that seems to be where it's actually used

kellybyrd replied:
done.

madninja commented on src/cmd/create.rs:
```diff
@@ -63,15 +73,16 @@ pub struct Sharded {
-    seed: bool,
+    /// Use a BIP39 or mobile app seed phrase to generate the wallet keys
+    /// (bip39, mobile)
+    seed: Option<SeedType>,
```
If you want to always define a type, use the structopt `default_value = "mobile"`. That way you don't need an Option here at all and avoid having expects or unwraps everywhere

kellybyrd replied:
But `--seed` is a optional on the cmdline. You can run `create` without it and the wallet just generates its own entropy.

madninja replied:
ah shoot, yeah, nvm this particular nit

madninja commented on src/mnemonic/mod.rs:
```diff
@@ -15,12 +19,26 @@ fn get_wordlist(language: Language) -> WordList {
+pub fn mnemonic_to_entropy(words: Vec<String>, seed_type: SeedType) -> Result<[u8; 32]> {
+    match seed_type {
+        SeedType::Bip39 => {
+            if words.len() != 12 && words.len() != 24 {
```
huh, so bip39 allows for 12 word seed strings? I wasn't aware of that at all

kellybyrd replied:
Heh. I thought the same thing. Its 12, 15, 18, 21, or 24. Any multiple of 3 from 12-24. AFAICT, 12 and 24 are by far the most common, and given the 256-bit entropy the wallet already wanted, easiest to implement.

madninja commented on src/mnemonic/mod.rs:
```diff
@@ -35,28 +53,69 @@ pub fn mnemonic_to_entropy(words: Vec<String>) -> Result<[u8; 32]> {
 }
 
+fn calc_checksum_128(bytes: [u8; 16]) -> u8 {
+    let mut hash = Sha256::new();
```
You already imported sha2::Digest, so use it? `Sha256::digest(bytes)` will make this one line

kellybyrd replied:
nice catch, thanks! I fixed this, but left these as functions rather than just using `Sha256::digest()` at the call-sites because I think it makes it easier to understand what is going on and maybe easier to add another checksum algo if there ends up being another SeedType.

madninja commented on src/mnemonic/mod.rs:
```diff
@@ -35,28 +53,69 @@ pub fn mnemonic_to_entropy(words: Vec<String>) -> Result<[u8; 32]> {
+}
+
+fn calc_checksum_256(bytes: [u8; 32]) -> u8 {
+    let mut hash = Sha256::new();
```
Same here

kellybyrd replied:
done.

madninja reviewed (commented):
This looks great! Just a few minor hygiene comments. 

I like the use of strum, I hadn't seen that yet. It adds a little bloat that could be avoided with simple FromStr and Display implementations, but I think that's ok

kellybyrd commented:
> This looks great! Just a few minor hygiene comments.
> 
> I like the use of strum, I hadn't seen that yet. It adds a little bloat that could be avoided with simple FromStr and Display implementations, but I think that's ok

I ended up copying the implementation of `Cmd::OutputFormat` and removing the need for strum.

madninja reviewed (changes requested):
The build fails cargo fmt.. can you fix?

kellybyrd commented:
> The build fails cargo fmt.. can you fix?

Yup. Fixed!
